eventauth/eventauth.go

```diff
@@ -362,3 +368,199 @@ func (e *eventAllower) commonChecks(event Event) error {
+	m.targetID = *event.StateKey
+	m.senderID = event.Sender
+	if m.create, err = newCreateContentFromAuthEvents(authEvents); err != nil {
+		return
+	}
+	if m.newMember, err = newMemberContentFromEvent(event); err != nil {
+		return
+	}
+	if m.oldMember, err = newMemberContentFromAuthEvents(authEvents, m.targetID); err != nil {
+		return
+	}
+	if m.senderMember, err = newMemberContentFromAuthEvents(authEvents, m.senderID); err != nil {
+		return
+	}
+	if m.powerLevels, err = newPowerLevelContentFromAuthEvents(authEvents, m.create.Creator); err != nil {
+		return
+	}
+	if m.newMember.Membership == "join" {
+		if m.joinRule, err = newJoinRuleContentFromAuthEvents(authEvents); err != nil {
+			return
+		}
+	}
+	return
+}
+
+// membershipAllowed checks whether the membership event is allowed
+func (m *membershipAllower) membershipAllowed(event Event) error {
+	if m.create.roomID != event.RoomID {
+		return errorf("create event has different roomID: %q != %q", event.RoomID, m.create.roomID)
+	}
+	if err := m.create.userIDAllowed(m.senderID); err != nil {
+		return err
+	}
+	if err := m.create.userIDAllowed(m.targetID); err != nil {
+		return err
+	}
+	// Special case the first join event in the room to allow the creator to join.
+	// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L328
+	if m.targetID == m.create.Creator &&
+		m.newMember.Membership == "join" &&
+		m.senderID == m.targetID &&
+		len(event.PrevEvents) == 1 {
+		// prev_events is a list of 2 element lists of event IDs and hashes.
+		if len(event.PrevEvents[0]) != 2 {
+			return errorf("unparsable prev event")
+		}
+		// Unmarshall the event ID string.
+		var prevEventID string
+		if err := json.Unmarshal(event.PrevEvents[0][0], &prevEventID); err != nil {
+			return errorf("unparsable prev event")
+		}
+		if prevEventID == m.create.eventID {
+			// If this is the room creator joining the room directly after the
+			// the create event, then allow.
+			return nil
+		}
+		// Otherwise fall back to the normal checks.
+	}
+
+	if m.newMember.Membership == "invite" && len(m.newMember.ThirdPartyInvite) != 0 {
+		// Special case third party invites
+		// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L393
+		panic(fmt.Errorf("ThirdPartyInvite not implemented"))
+	}
+
+	if m.targetID == m.senderID {
+		// If the state_key and the sender are the same then this is an attempt
+		// by a user to update their own membership.
+		return m.membershipAllowedSelf()
+	}
+	// Otherwise this is an attempt to modify the membership of somebody else.
+	return m.membershipAllowedOther()
+}
+
+// membershipAllowedSelf determines if the change made by the user to their own membership is allowed.
+func (m *membershipAllower) membershipAllowedSelf() error {
+	if m.newMember.Membership == "join" {
+		// A user that is not in the room is allowed to join if the room
+		// join rules are "public".
+		if m.oldMember.Membership == "leave" && m.joinRule.JoinRule == "public" {
```

Comment:
I'd appreciate it if you could yank `"leave"` and co out to be constants so the compiler will whine if we mis-spell.